src/matrixrtc/MatrixRTCSession.ts

```diff
@@ -0,0 +1,352 @@
+        for (const memberEvent of callMemberEvents) {
+            const eventMemberships: CallMembershipData[] = memberEvent.getContent()["memberships"];
+            if (eventMemberships === undefined) {
+                logger.warn("Ignoring malformed member event: no memberships section");
+                continue;
+            }
+            if (!Array.isArray(eventMemberships)) {
+                logger.warn("Malformed member event: memberships is not an array");
+                continue;
+            }
+
+            for (const membershipData of eventMemberships) {
+                try {
+                    const membership = new CallMembership(memberEvent, membershipData);
+
+                    if (membership.callId !== "" || membership.scope !== "m.room") {
+                        // for now, just ignore anything that isn't the a room scope call
+                        logger.info(`Ignoring user-scoped call`);
+                        continue;
+                    }
+
+                    if (membership.isExpired()) {
+                        logger.info(
+                            `Ignoring expired device membership ${memberEvent.getSender()}/${membership.deviceId}`,
+                        );
+                        continue;
+                    }
+                    callMemberships.push(membership);
+                } catch (e) {
+                    logger.warn("Couldn't construct call membership: ", e);
+                }
+            }
+        }
+
+        callMemberships.sort((a, b) => a.createdTs() - b.createdTs());
+
+        return callMemberships;
+    }
+
+    /**
+     * Return a the MatrixRTC for the room, whether there are currently active members or not
+     */
+    public static roomSessionForRoom(client: MatrixClient, room: Room): MatrixRTCSession {
+        const callMemberships = MatrixRTCSession.callMembershipsForRoom(room);
+
+        return new MatrixRTCSession(client, room, callMemberships);
+    }
+
+    private constructor(
+        private readonly client: MatrixClient,
+        public readonly room: Room,
+        public memberships: CallMembership[],
+    ) {
+        super();
+        this.setExpiryTimer();
+    }
+
+    public isJoined(): boolean {
+        return this.relativeExpiry !== undefined;
+    }
+
+    /**
+     * Performs cleanup & removes timers for client shutdown
+     */
+    public stop(): void {
+        this.leaveRoomSession();
+        if (this.expiryTimeout) {
+            clearTimeout(this.expiryTimeout);
+            this.expiryTimeout = undefined;
+        }
+    }
+
+    /**
+     * Announces this user and device as joined to the MatrixRTC session,
+     * and continues to update the membership event to keep it valid until
+     * leaveRoomSession() is called
+     * This will not subscribe to updates: remember to call subscribe() separately if
+     * desired.
+     */
+    public joinRoomSession(activeFoci: Focus[]): void {
```

Comment:
Should we handle local echo in here somehow?

Reply:
I had been wondering about this, but as yet nothing has needed it. I'm sort of inclined to say that the local client's idea of whether they're joined or not should just depend on whether they think they ought to be (is. `isJoined()`) which is effectively local echo but way simpler. Wdyt?

Reply:
Yeah, I guess it really depends on whether or not any of the UI depends on the `memberships`

Reply:
Yeah, if at some point it does then we can make it support local echo